### PR TITLE
Notes & Questions for PR 17428

### PR DESCRIPTION
**Pull request by amitiuttarwar**
First pass at updating notes & questions. Review is definitely needed for language & formatting.

### CODE REVIEW

File: _posts/2020-02-26-#17428.md [hunk 1]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+
+- In the 2015 implementation of Bitcoin, an attacker could carry out a
+  "connection starvation attack" where the attacker forces a victim to reboot and fills up all the free
+  connection slots in the network, disabling the victim from connecting to any
+  nodes. So, anchor connections would not have been an effective mitigation
+  strategy.
+
+- Since then, Bitcoin has implemented logic to sometimes rotate through
+  incoming connections when the max number is hit. This means the victim has a
+  way to successfully evict attackers connections and reconnect to its peers.
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: nit: s/attackers/the attacker's/

File: _posts/2020-02-26-#17428.md [hunk 2]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+  While this means its possible for an attacker to eclipse a node without it
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/its/it's/

File: _posts/2020-02-26-#17428.md [hunk 3]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+  rebooting, it also makes carrying out the eclipse attack more complex.
+
+- Reconnecting outgoing connections makes the network graph more static. While
+  this helps maintain longstanding honest connections and prevent an eclipse
+  attack, this can have a negative effect on privacy since the network topology
+  is easier to map.
+
+- Other reasons anchoring all connections would be bad includes
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/includes/include:/

File: _posts/2020-02-26-#17428.md [hunk 4]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+1. A victim of an eclipse attack has no way of escaping
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/escaping/escaping./ (or remove the periods from points 2 and 3)

File: _posts/2020-02-26-#17428.md [hunk 5]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+2. Strong persistence can contribute to the network self-partitioning. For example, if the longer
+  distance connections are less reliable, nodes are incentivized to connect
+  locally and this could lead to subgraphs per continent.
+3. An attacker could eclipse nodes with a small number of nodes but 100% uptime
+  and a large capacity for inbound connections. Every time a node adds a
+  connection, if its an adversary that malicious node would be "locked in"
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/its/it's/

File: _posts/2020-02-26-#17428.md [hunk 6]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+- The 2 outbound block-relay-only connections are written to the `anchors.dat`
+  file. When a node restarts, it first tries establishing connections with the
+  anchors before attempting to connect to other peers.
+
+- Previously, a node that would quickly reattempt the same outgoing connections
+  was detected as a spy or mass connector. This new behavior means that honest
+  nodes will attempt the same outgoing connections more than once, which was
+  unlikely before.
+
+- The current state of this implementation introduces a new risk- if one of the
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/risk- if/risk -- if/

File: _posts/2020-02-26-#17428.md [hunk 7]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+  anchor peers is malicious and is able to exploit a DoS vulnerability, they
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/they/it/

File: _posts/2020-02-26-#17428.md [hunk 8]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+  can repeatedly crash the victim's node.
+
+  The proposed fix is to forget the anchor peers on an unclean exit.
+
+  In reaction to these findings, the PR was [closed](https://github.com/bitcoin/bitcoin/pull/17428#issuecomment-583737058)
+  but then re-opened based on a comment saying there are still advantages of
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: s/of/to/ and perhaps "comment by Gregory Maxwell"

File: _posts/2020-02-26-#17428.md [hunk 9]
```diff
@@ -44,9 +50,92 @@ _Preliminary notes below. Additional notes will be published on Monday, February
...
+  a. When are peers persisted and removed?
+  b. When does the node ignore the contents of anchors.dat?
+  c. What are some implications of the timing of adding peers?
+
+4. What are the conditions for adding an anchor?
+  a. Why are the anchors blocks-only peers?
+  b. What would be the issue of using normal (transaction & block relay) peers
+  as anchors?
+  c. What are limitations of how this implementation chooses anchors?
+  d. Why limit to only 2 anchors?
```
Comment (jonatack) [inline _posts/2020-02-26-#17428.md]: Questions 2, 3 and 4 might be more readable as bullet points or numbered markdown points.

**Comment by jnewbery** [comment]
I'll fix the style nits in a follow-up

**Pull request merged**